[WebIDL] Eagerly allocate element wrappers onto a butterfly storage of a static NodeList

shvaikalesh opened the pull request:
#### e0c79898e6c5e72041f0228bc6378027b7fd8ff2
<pre>
[WebIDL] Eagerly allocate element wrappers onto a butterfly storage of a static NodeList
<a href="https://bugs.webkit.org/show_bug.cgi?id=234538">https://bugs.webkit.org/show_bug.cgi?id=234538</a>

Reviewed by Yusuke Suzuki and Saam Barati.

This change greatly speeds up iterating static NodeList iteration (including via
Array.prototype.indexOf), making it on par with JSArray. It preserves current
(weird) NodeList semantics, including cases of transitioning to slow put array storage
mode, and is future-proof against aligning internal methods of DOM collections with
other browsers (<a href="http://webkit.org/b/218849).">http://webkit.org/b/218849).</a>

There was no way to achieve that with existing indexing types: put_by_val compiled
for NonArrayWithContiguous emits only indexing type check, thus incorrectly mutating
a NodeList instead of calling put() override. NonArrayWithSlowPutArrayStorage could
be semantically correct alternative only if all its elements are onto a sparse map,
which won&apos;t be much faster.

Two viable approaches were either introducing a new Contiguous-like indexing type,
or doing something similar to IndexedDirectArgumentsLoad. This change does the former
to reuse existing logic for DFG::ArrayMode refining and ArrayIndexOf compilation.

Another tricky part is that NodeList&apos;s &quot;length&quot; is a configurable getter on prototype,
unlike own non-configurable JSArray&apos;s &quot;length&quot;, and we need to ensure that both the
NodeList structure is original (inherits from NodeList.prototype) and the &quot;length&quot;
getter wasn&apos;t deleted / overriden.

For the static NodeList to leverage fast compiled get_by_val, it needs to have original
getOwnPropertySlotByIndex() method, so SlowPutArrayStorageVectorPropertiesAreReadOnly
out-of-line type info flag is added to reliably preserve ReadOnly attributes of its
indices as per spec [1].

This change is a locally confirmed 2% progression on all three Speedometer2/VanillaJS subtests.

[1] <a href="https://webidl.spec.whatwg.org/#LegacyPlatformObjectGetOwnProperty">https://webidl.spec.whatwg.org/#LegacyPlatformObjectGetOwnProperty</a> (step 1.2.7)

* .gitignore:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/AdaptiveInferredPropertyValueWatchpointBase.cpp:
* Source/JavaScriptCore/bytecode/AdaptiveInferredPropertyValueWatchpointBase.h:
* Source/JavaScriptCore/bytecode/ArrayProfile.cpp:
(JSC::dumpArrayModes):
(JSC::ArrayProfile::computeUpdatedPrediction):
* Source/JavaScriptCore/bytecode/ArrayProfile.h:
(JSC::shouldUseAlwaysSlowPutContiguous):
* Source/JavaScriptCore/bytecode/PolymorphicAccess.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
(JSC::DFG::ArrayMode::fromObserved):
(JSC::DFG::ArrayMode::refine const):
(JSC::DFG::ArrayMode::alreadyChecked const):
(JSC::DFG::arrayTypeToString):
(JSC::DFG::toIndexingShape):
(JSC::DFG::permitsBoundsCheckLowering):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::usesButterfly const):
(JSC::DFG::ArrayMode::isSlowPut const):
(JSC::DFG::ArrayMode::isAnyKindOfContiguous const):
(JSC::DFG::ArrayMode::lengthNeedsStorage const):
(JSC::DFG::ArrayMode::modeForPut const):
(JSC::DFG::ArrayMode::supportsSelfLength const):
(JSC::DFG::ArrayMode::arrayModesThatPassFiltering const):
* Source/JavaScriptCore/dfg/DFGArrayifySlowPathGenerator.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::attemptToMakeGetArrayLength):
(JSC::DFG::FixupPhase::attemptToMakeGetArrayLengthForAlwaysSlowPutContiguous):
(JSC::DFG::FixupPhase::fixupArrayIndexOf):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::jumpSlowForUnwantedArrayMode):
(JSC::DFG::SpeculativeJIT::checkArray):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
(JSC::FTL::AbstractHeapRepository::forArrayType):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayify):
(JSC::FTL::DFG::LowerDFGToB3::compileGetArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOf):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::performLLIntGetByID):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::tryFastIndexOf):
(JSC::fastIndexOf): Deleted.
* Source/JavaScriptCore/runtime/IndexingType.cpp:
(JSC::dumpIndexingType):
* Source/JavaScriptCore/runtime/IndexingType.h:
(JSC::hasAlwaysSlowPutContiguous):
(JSC::hasAnyContiguous):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastSlice):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::initAlwaysSlowPutContiguousWatchpointSet):
(JSC::JSGlobalObject::recordOriginalAlwaysSlowPutContiguousStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::alwaysSlowPutContiguousWatchpointSet):
(JSC::JSGlobalObject::isOriginalSlowPutContigiousStructure):
(JSC::JSGlobalObject::originalAlwaysSlowPutContiguousStructures const):
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::alwaysSlowPutContiguousPrototypeChainIsSane):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::getOwnPropertySlotByIndex):
(JSC::JSObject::convertContiguousToArrayStorage):
(JSC::JSObject::ensureArrayStorageExistsAndEnterDictionaryIndexingMode):
(JSC::JSObject::switchToSlowPutArrayStorage):
(JSC::JSObject::putByIndexBeyondVectorLengthWithoutAttributes):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::trySetIndexQuickly):
(JSC::JSObject::setIndexQuickly):
(JSC::JSObject::tryMakeWritableContiguous):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::canDoFastIndexedAccess):
* Source/JavaScriptCore/runtime/JSTypeInfo.h:
(JSC::TypeInfo::arrayStorageVectorPropertiesAreReadOnly const):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::newIndexingType):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::createAlwaysSlowPutContiguousObject):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):
* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::every const):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::finishCreation):
(WebCore::JSDOMWindowBase::installAlwaysSlowPutContiguousWatchpoints):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSNodeListCustom.cpp:
(WebCore::JSStaticNodeList::createPrototype):
(WebCore::JSStaticNodeList::JSStaticNodeList):
(WebCore::JSStaticNodeList::finishCreation):
(WebCore::createWrapper):
* Source/WebCore/bindings/js/JSNodeListCustom.h:
(WebCore::JSStaticNodeList::create):
(WebCore::JSStaticNodeList::createStructure):
* Source/WebCore/dom/NodeList.h:
(WebCore::NodeList::isStaticNodeList const):
* Source/WebCore/dom/StaticNodeList.h:

Canonical link: <a href="https://commits.webkit.org/253834@main">https://commits.webkit.org/253834@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05f2d782b44efa1eca7d26970aed178c2eb860b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87202 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31289 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18030 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96233 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149786 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29653 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91214 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92818 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23956 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74007 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78933 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79177 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79065 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27378 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13014 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72741 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27325 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14029 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25916 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29010 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75553 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1089 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28949 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33306 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16728 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/bindings/js/JSNodeListCustom.cpp:
```diff
@@ -37,6 +37,80 @@
+        &JSStaticNodeList::visitChildren,
+        &JSStaticNodeList::visitOutputConstraints,
+        &JSStaticNodeList::visitOutputConstraints,
+    },
```
CREATE_METHOD_TABLE should work? If not, probably we should rearchitect the design of hierarchy of `JSNodeList` and `JSStaticNodeList`. (inserting `JSNodeListBase` which implements shared methods for both).

shvaikalesh replied:
`CREATE_METHOD_TABLE` doesn't work as it expands to `&JSStaticNodeList::foo` for all methods, yet we need original `getOwnPropertySlot` and friends to be fast.

Rearchitecting inheritance chain into `JSNodeListBase` / `JSStaticNodeList` / `JSLiveNodeList` is indeed what we should strive for, but currently that would amount to massive changes of `CodeGeneratorJS.pm`, unfortunately.

Constellation replied:
Let's add FIXME since it is not ideal.

shvaikalesh replied:
Added.

Constellation commented on Source/WebCore/bindings/js/JSNodeListCustom.h:
```diff
@@ -26,9 +26,40 @@
 
 namespace WebCore {
 
+class JSStaticNodeList : public JSNodeList {
```
Put `final`

shvaikalesh replied:
Done, thanks.

Constellation reviewed (commented):
quick comments.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -898,6 +904,14 @@ class JSGlobalObject : public JSSegmentedVariableObject {
+    {
+        ASSERT(hasAlwaysSlowPutContiguous(structure->indexingMode()));
+        return m_originalAlwaysSlowPutContiguousStructures.contains(structure->id());
+    }
```
Instead of this, how about checking `didTransition` bit of this structure? Does that work?

shvaikalesh replied:
This bit was re-written into having `JSGlobalObject::m_originalAlwaysSlowPutContiguousStructureSet`, which we either way need to emit `CheckStructure`.

shvaikalesh replied:
`didTransition` + `indexingMode` check would indeed worked out there.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -8644,7 +8652,9 @@ void SpeculativeJIT::compileGetArrayLength(Node* node)
-    case Array::Contiguous: {
+    case Array::Contiguous:
+    case Array::AlwaysSlowPutContiguous: {
+        dataLogLn("compileGetArrayLength");
```
`dataLogLn` is remaining.

shvaikalesh replied:
Removed, sorry.

Constellation commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -10075,6 +10085,7 @@ void SpeculativeJIT::compileArraySlice(Node* node)
 
 void SpeculativeJIT::compileArrayIndexOf(Node* node)
 {
+    dataLogLn("compileArrayIndexOf");
```
Ditto

shvaikalesh replied:
Removed.

Constellation commented on Source/JavaScriptCore/bytecode/AdaptiveInferredPropertyValueWatchpointBase.h:
```diff
@@ -34,7 +34,7 @@ namespace JSC {
 // FIXME: This isn't actually a Watchpoint. We should probably have a name which better reflects that:
 // https://bugs.webkit.org/show_bug.cgi?id=202381
-class AdaptiveInferredPropertyValueWatchpointBase {
+class JS_EXPORT_PRIVATE AdaptiveInferredPropertyValueWatchpointBase {
```
Let's avoid annotating the entire class with `JS_EXPORT_PRIVATE`. And let's annotate each functions.

shvaikalesh replied:
Done.

Constellation commented on Source/JavaScriptCore/bytecode/Watchpoint.h:
```diff
@@ -42,16 +42,16 @@ struct ArrayBufferViewWatchpointAdaptor;
 class VM;
 
-class FireDetail {
+class JS_EXPORT_PRIVATE FireDetail {
```
Let's avoid annotating the entire class with JS_EXPORT_PRIVATE. And let's annotate each functions.
Historically, annotating the entire class caused many problems.

shvaikalesh replied:
Fixed.

Constellation commented on Source/JavaScriptCore/bytecode/Watchpoint.h:
```diff
@@ -42,16 +42,16 @@ struct ArrayBufferViewWatchpointAdaptor;
 };
 
-class StringFireDetail final : public FireDetail {
+class JS_EXPORT_PRIVATE StringFireDetail final : public FireDetail {
```
Ditto.

shvaikalesh replied:
Done, thanks.

Constellation commented on Source/WebCore/bindings/js/JSNodeListCustom.cpp:
```diff
@@ -37,6 +37,79 @@
+    auto contiguousValues = butterfly->contiguous();
+
+    for (unsigned index = 0; index < length; index++) {
+        auto jsNode = toJS(globalObject, globalObject, *nodeList.item(index));
```
Should we insert error check?

saambarati replied:
What's keeping these values alive as we call toJS? Can't toJS allocate?

shvaikalesh replied:
No error check needed, this `toJS` can't possibly throw.

I've switched to setting a butterfly with cleared values before this loop. I hope that won't be a noticeable regression.

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -559,8 +559,8 @@ const NoIndexingShape          = constexpr NoIndexingShape
-const SlowPutArrayStorageShape = constexpr SlowPutArrayStorageShape
 const CopyOnWrite              = constexpr CopyOnWrite
 const ArrayWithUndecided       = constexpr ArrayWithUndecided
 
```
Let's adjust `=` place of all shapes.

shvaikalesh replied:
Adjusted here and in `IndexingType.h`.

Constellation commented on Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
```diff
@@ -134,6 +135,37 @@ void JSDOMWindowBase::finishCreation(VM& vm, JSWindowProxy* proxy)
+
+void JSDOMWindowBase::installAlwaysSlowPutContiguousWatchpoint(VM& vm)
+{
+    DeferTermination deferScope(vm);
```
Let's use DeferTerminationForAWhile.

shvaikalesh replied:
Done.

Constellation commented on Source/JavaScriptCore/runtime/ObjectPropertyChangeAdaptiveWatchpoint.h:
```diff
@@ -30,7 +30,7 @@
 
 template<typename WatchpointSet>
-class ObjectPropertyChangeAdaptiveWatchpoint final : public AdaptiveInferredPropertyValueWatchpointBase {
+class JS_EXPORT_PRIVATE ObjectPropertyChangeAdaptiveWatchpoint final : public AdaptiveInferredPropertyValueWatchpointBase {
```
Avoid using `JS_EXPORT_PRIVATE` for entire class.

Constellation commented on Source/JavaScriptCore/runtime/JSObject.cpp:
```diff
@@ -572,6 +572,8 @@ bool JSObject::getOwnPropertySlotByIndex(JSObject* thisObject, JSGlobalObject* g
     if (i > MAX_ARRAY_INDEX)
         return thisObject->methodTable()->getOwnPropertySlot(thisObject, globalObject, Identifier::from(vm, i), slot);
+
+    unsigned attributes = thisObject->structure()->typeInfo().arrayStorageVectorPropertiesAreReadOnly() ? static_cast<unsigned>(PropertyAttribute::ReadOnly) : static_cast<unsigned>(PropertyAttribute::None);
```
Can you avoid `thisObject->structure()->typeInfo().arrayStorageVectorPropertiesAreReadOnly()` for most of indexing types? We should not load structure if it is not necessary. And I think most of indexing type never gets `arrayStorageVectorPropertiesAreReadOnly` = true. (Only slow-put ones)

saambarati replied:
Yeah, probably we can just add a helper onto JSObject that does the fast thing depending on indexing type

shvaikalesh replied:
I've made the structure's `typeInfo()` check conditional based on indexing type, only for slow-puts. I'm not sure that at this point (only one usage) it's worth to extract a helper though.

Constellation reviewed (approved):
Can you add detailed description of this change in the commit message?

saambarati commented on Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
```diff
@@ -4330,6 +4334,7 @@ bool AbstractInterpreter<AbstractStateType>::executeEffects(unsigned clobberLimi
             case Array::Int32:
             case Array::Double:
             case Array::Contiguous:
+            case Array::AlwaysSlowPutContiguous:
```
do we have tests for `for in`?

shvaikalesh replied:
Added `JSTests/stress/for-in-always-slow-put-contiguous.js`.

saambarati commented on Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
```diff
@@ -461,7 +470,17 @@ bool ArrayMode::alreadyChecked(Graph& graph, Node* node, const AbstractValue& va
     // But here, we already filtered TypedArrays. So, just handle it like a NonArray.
-    case Array::OriginalNonArray:
+    case Array::OriginalNonArray: {
+        if (type() == Array::AlwaysSlowPutContiguous) {
+            if (!value.m_structure.isFinite())
+                return false;
+            JSGlobalObject* globalObject = graph.globalObjectFor(node->origin.semantic);
+            return value.m_structure.toStructureSet().isSubsetOf(globalObject->originalAlwaysSlowPutContiguousStructureSet());
+        }
```
Should we call the other helper here?

shvaikalesh replied:
@saambarati Could please expand on "the other helper" here? Thanks.

saambarati replied:
Not sure what I was thinking. Let's ignore this.

saambarati commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -959,7 +961,7 @@ class JSObject : public JSCell {
+        if (LIKELY(hasAnyContiguous(indexingType()) && !isCopyOnWrite(indexingMode())))
             return m_butterfly->contiguous();
             
         return tryMakeWritableContiguousSlow(vm);
```
Is this doing what we expect in arrayify?

shvaikalesh replied:
That is a really nice catch, thank you. Fixed `tryMakeWritableContiguous` to return an empty values for new indexing type.

saambarati commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -532,6 +532,7 @@ class JSGlobalObject : public JSSegmentedVariableObject {
     InlineWatchpointSet m_sharedArrayBufferSpeciesWatchpointSet { ClearWatchpoint };
     InlineWatchpointSet m_typedArrayConstructorSpeciesWatchpointSet { IsWatched };
     InlineWatchpointSet m_typedArrayPrototypeIteratorProtocolWatchpointSet { IsWatched };
+    InlineWatchpointSet m_alwaysSlowPutContiguousWatchpointSet { ClearWatchpoint };
```
We should give this a more descriptive name. I think it's even ok if it refers to the DOM in its name. But right now, this name isn't describing what's being protected. Might also be worth a comment saying how we can rely on this watchpoint. For example, it's protecting something on the prototype. And it's protecting the "length" getter on the prototype. We still need to protect against self length properties. And we need to ensure self's prototype is the prototype this watchpoint is installed on. Generally, we do that via an original structure check

shvaikalesh replied:
Renamed to `m_alwaysSlowPutContiguousPrototypesAreSaneWatchpointSet` and added a comment on `NodeList.prototype.length`.

We do perform original structure check already, and I'm adding a thorough WPT coverage for that.

saambarati commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -4117,13 +4117,39 @@ class FixupPhase : public Phase {
+        }
+
+        return false;
+    }
```
We should only do this if profiling tells us we're the original structure, right? Otherwise, this is an OSR exit loop AFAICT

saambarati replied:
Ignore me, I see that we're relying on profiling above that we're "original"

saambarati commented on Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
```diff
@@ -863,6 +863,13 @@ static JSValue performLLIntGetByID(BytecodeIndex bytecodeIndex, CodeBlock* codeB
 
             if (!(--metadata.hitCountForLLIntCaching))
                 setupGetByIdPrototypeCache(globalObject, vm, codeBlock, bytecodeIndex, metadata, baseCell, slot, ident);
+        } else if (hasAlwaysSlowPutContiguous(baseCell->indexingMode()) && ident == vm.propertyNames->length) {
```
You want Options::useLLIntICs here, too

shvaikalesh replied:
Please note that a parent branch is guarded by a `Options::useLLIntICs` on L816.

saambarati commented on Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
```diff
@@ -1747,8 +1747,8 @@ llintOpWithMetadata(op_get_by_val, OpGetByVal, macro (size, get, dispatch, metad
 .opGetByValNotDouble:
-    subi ArrayStorageShape, t2
-    bia t2, SlowPutArrayStorageShape - ArrayStorageShape, .opGetByValNotIndexedStorage
+    bieq t2, AlwaysSlowPutContiguousShape, .opGetByValIsContiguous
+    bilt t2, ArrayStorageShape, .opGetByValNotIndexedStorage
```
Why does the old code not work?

shvaikalesh replied:
It does work correctly, but takes a slow path for new indexing shape. The proposed change leverages the fact that `SlowPutArrayStorageShape` is the last indexing shape (arithmetically) to maintain the same # of opcodes while ensuring `AlwaysSlowPutContiguousShape` is fast even in LLInt.

saambarati commented on Source/WebCore/bindings/js/JSNodeListCustom.cpp:
```diff
@@ -37,6 +37,84 @@
+
+    for (unsigned i = 0; i < length; ++i) {
+        auto jsNode = toJS(globalObject, globalObject, *nodeList.item(i));
+        contiguousValues.atUnsafe(i).setWithoutWriteBarrier(jsNode);
```
We need to be emitting a barrier here since toJS can allocate.

shvaikalesh replied:
Replaced with `.set(`, thanks!

shvaikalesh committed:
[WebIDL] Eagerly allocate element wrappers onto a butterfly storage of a static NodeList
https://bugs.webkit.org/show_bug.cgi?id=234538

Reviewed by Yusuke Suzuki and Saam Barati.

This change greatly speeds up iterating static NodeList iteration (including via
Array.prototype.indexOf), making it on par with JSArray. It preserves current
(weird) NodeList semantics, including cases of transitioning to slow put array storage
mode, and is future-proof against aligning internal methods of DOM collections with
other browsers (http://webkit.org/b/218849).

There was no way to achieve that with existing indexing types: put_by_val compiled
for NonArrayWithContiguous emits only indexing type check, thus incorrectly mutating
a NodeList instead of calling put() override. NonArrayWithSlowPutArrayStorage could
be semantically correct alternative only if all its elements are onto a sparse map,
which won't be much faster.

Two viable approaches were either introducing a new Contiguous-like indexing type,
or doing something similar to IndexedDirectArgumentsLoad. This change does the former
to reuse existing logic for DFG::ArrayMode refining and ArrayIndexOf compilation.

Another tricky part is that NodeList's "length" is a configurable getter on prototype,
unlike own non-configurable JSArray's "length", and we need to ensure that both the
NodeList structure is original (inherits from NodeList.prototype) and the "length"
getter wasn't deleted / overriden.

For the static NodeList to leverage fast compiled get_by_val, it needs to have original
getOwnPropertySlotByIndex() method, so SlowPutArrayStorageVectorPropertiesAreReadOnly
out-of-line type info flag is added to reliably preserve ReadOnly attributes of its
indices as per spec [1].

This change is a locally confirmed 2% progression on all three Speedometer2/VanillaJS subtests.

[1] https://webidl.spec.whatwg.org/#LegacyPlatformObjectGetOwnProperty (step 1.2.7)

* .gitignore:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::generateWithGuard):
(JSC::AccessCase::generateImpl):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/AdaptiveInferredPropertyValueWatchpointBase.cpp:
* Source/JavaScriptCore/bytecode/AdaptiveInferredPropertyValueWatchpointBase.h:
* Source/JavaScriptCore/bytecode/ArrayProfile.cpp:
(JSC::dumpArrayModes):
(JSC::ArrayProfile::computeUpdatedPrediction):
* Source/JavaScriptCore/bytecode/ArrayProfile.h:
(JSC::shouldUseAlwaysSlowPutContiguous):
* Source/JavaScriptCore/bytecode/PolymorphicAccess.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):
* Source/JavaScriptCore/bytecode/Watchpoint.h:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGArrayMode.cpp:
(JSC::DFG::ArrayMode::fromObserved):
(JSC::DFG::ArrayMode::refine const):
(JSC::DFG::ArrayMode::alreadyChecked const):
(JSC::DFG::arrayTypeToString):
(JSC::DFG::toIndexingShape):
(JSC::DFG::permitsBoundsCheckLowering):
* Source/JavaScriptCore/dfg/DFGArrayMode.h:
(JSC::DFG::ArrayMode::usesButterfly const):
(JSC::DFG::ArrayMode::isSlowPut const):
(JSC::DFG::ArrayMode::isAnyKindOfContiguous const):
(JSC::DFG::ArrayMode::lengthNeedsStorage const):
(JSC::DFG::ArrayMode::modeForPut const):
(JSC::DFG::ArrayMode::supportsSelfLength const):
(JSC::DFG::ArrayMode::arrayModesThatPassFiltering const):
* Source/JavaScriptCore/dfg/DFGArrayifySlowPathGenerator.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::attemptToMakeGetArrayLength):
(JSC::DFG::FixupPhase::attemptToMakeGetArrayLengthForAlwaysSlowPutContiguous):
(JSC::DFG::FixupPhase::fixupArrayIndexOf):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
(JSC::DFG::SpeculativeJIT::jumpSlowForUnwantedArrayMode):
(JSC::DFG::SpeculativeJIT::checkArray):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
* Source/JavaScriptCore/ftl/FTLAbstractHeapRepository.h:
(JSC::FTL::AbstractHeapRepository::forArrayType):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayify):
(JSC::FTL::DFG::LowerDFGToB3::compileGetArrayLength):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOf):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::performLLIntGetByID):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::tryFastIndexOf):
(JSC::fastIndexOf): Deleted.
* Source/JavaScriptCore/runtime/IndexingType.cpp:
(JSC::dumpIndexingType):
* Source/JavaScriptCore/runtime/IndexingType.h:
(JSC::hasAlwaysSlowPutContiguous):
(JSC::hasAnyContiguous):
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::fastSlice):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::initAlwaysSlowPutContiguousWatchpointSet):
(JSC::JSGlobalObject::recordOriginalAlwaysSlowPutContiguousStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::alwaysSlowPutContiguousWatchpointSet):
(JSC::JSGlobalObject::isOriginalSlowPutContigiousStructure):
(JSC::JSGlobalObject::originalAlwaysSlowPutContiguousStructures const):
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::alwaysSlowPutContiguousPrototypeChainIsSane):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::getOwnPropertySlotByIndex):
(JSC::JSObject::convertContiguousToArrayStorage):
(JSC::JSObject::ensureArrayStorageExistsAndEnterDictionaryIndexingMode):
(JSC::JSObject::switchToSlowPutArrayStorage):
(JSC::JSObject::putByIndexBeyondVectorLengthWithoutAttributes):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::JSObject::trySetIndexQuickly):
(JSC::JSObject::setIndexQuickly):
(JSC::JSObject::tryMakeWritableContiguous):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::canDoFastIndexedAccess):
* Source/JavaScriptCore/runtime/JSTypeInfo.h:
(JSC::TypeInfo::arrayStorageVectorPropertiesAreReadOnly const):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::newIndexingType):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::createAlwaysSlowPutContiguousObject):
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):
* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::every const):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::JSDOMWindowBase::finishCreation):
(WebCore::JSDOMWindowBase::installAlwaysSlowPutContiguousWatchpoints):
* Source/WebCore/bindings/js/JSDOMWindowBase.h:
* Source/WebCore/bindings/js/JSNodeListCustom.cpp:
(WebCore::JSStaticNodeList::createPrototype):
(WebCore::JSStaticNodeList::JSStaticNodeList):
(WebCore::JSStaticNodeList::finishCreation):
(WebCore::createWrapper):
* Source/WebCore/bindings/js/JSNodeListCustom.h:
(WebCore::JSStaticNodeList::create):
(WebCore::JSStaticNodeList::createStructure):
* Source/WebCore/dom/NodeList.h:
(WebCore::NodeList::isStaticNodeList const):
* Source/WebCore/dom/StaticNodeList.h:

Canonical link: https://commits.webkit.org/253834@main
